buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+ * <p>
+ * The build know the current version, and we parse server code to find the rest making sure that these match.
+ * We can reliably figure out which the unreleased versions are due to the convention of always adding the next unreleased
+ * version number to server in all branches when a version is released.
+ * This convention is enforced by checking the versions we consider to be unreleased against an
+ * authoritative source (maven central).
+ */
+public class VersionCollection {
+
+    private static final Pattern LINE_PATTERN = Pattern.compile(
+        "\\W+public static final Version V_(\\d+)_(\\d+)_(\\d+)(_alpha\\d+|_beta\\d+|_rc\\d+)? .*"
+    );
+
+    private final Version currentVersion;
+    private final Map<Integer, List<Version>> groupByMajor;
+
+    public class UnreleasedVersionDescription {
+        private final Version version;
+        private final String branch;
+        private final String gradleProjectName;
+
+        UnreleasedVersionDescription(Version version, String branch, String gradleProjectName) {
+            this.version = version;
+            this.branch = branch;
+            this.gradleProjectName = gradleProjectName;
+        }
+
+        public Version getVersion() {
+            return version;
+        }
+
+        public String getBranch() {
+            return branch;
+        }
+
+        public String getGradleProjectName() {
+            return gradleProjectName;
+        }
+    }
+
+    public VersionCollection(List<String> versionLines) {
+        this(versionLines, VersionProperties.getElasticsearch());
+    }
+
+    protected VersionCollection(List<String> versionLines, Version currentVersionProperty) {
+        groupByMajor = versionLines.stream()
+            .map(LINE_PATTERN::matcher)
+            .filter(Matcher::matches)
+            .map(match -> new Version(
+                Integer.parseInt(match.group(1)),
+                Integer.parseInt(match.group(2)),
+                Integer.parseInt(match.group(3)),
+                (match.group(4) == null ? "" : match.group(4)).replace('_', '-'),
+                false
+            ))
+            .filter(version -> version.getSuffix().isEmpty() || version.equals(currentVersionProperty))
+            .collect(Collectors.groupingBy(Version::getMajor, Collectors.toList()));
+
+        if (groupByMajor.isEmpty()) {
+            throw new IllegalArgumentException("Could not parse any versions");
+        }
+
+        currentVersion = getLatestVersionByKey(
+            groupByMajor,
+            groupByMajor.keySet().stream().max(Integer::compareTo)
+                .orElseThrow(() -> new IllegalStateException("Unexpected number of versions in collection"))
+        );
+
+        assertCurrentVersionMatchesParsed(currentVersionProperty);
+
+        assertNoOlderThanTwoMajors();
+
+        markUnreleasedAsSnapshot();
+    }
+
+    private void markUnreleasedAsSnapshot() {
+        getUnreleased()
+            .forEach(unreleased -> {
```

Comment:
nit: this extra line adds horizontal block spacing without any benefit I can see. It could be on the same line?

```
getUnreleased().forEach(unreleased -> {
```